houston/sandbox.py

```diff
@@ -26,193 +25,152 @@ class Sandbox(object):
     executing test cases on a given system.
     """
     def __init__(self,
-                 system: 'System',
-                 client_bugzoo: BugZooClient
+                 client_bugzoo: BugZooClient,
+                 container: Container,
+                 state_initial: State,
+                 environment: Environment,
+                 configuration: Configuration
                  ) -> None:
         self.__lock = threading.Lock()
-        self.__system = system
-        self.__snapshot = system.snapshot
         self._bugzoo = client_bugzoo
-        self.__container = client_bugzoo.containers.provision(self.__snapshot)
-        self.__instrumented = False
+        self.__container = container
+        self.__state = state_initial
+        self.__environment = environment
+        self.__configuration = configuration
 
     @property
-    def snapshot(self) -> Snapshot:
+    def running_time(self) -> float:
         """
-        A BugZoo snapshot of the system under test.
+        Returns the number of seconds (wall-clock time) that have elapsed
+        since this sandbox session begun.
         """
-        return self.__snapshot
+        raise NotImplementedError
 
     @property
-    def system(self) -> 'System':
+    def state(self) -> State:
         """
-        The system under test by this sandbox.
+        The last observed state of the system under test.
         """
-        return self.__system
-
-    sut = system
+        return self.__state
 
     @property
-    def container(self) -> Optional[bugzoo.Container]:
+    def configuration(self) -> Configuration:
         """
-        The BugZoo container underlying this sandbox.
+        The configuration used by the system under test.
         """
-        return self.__container
+        return self.__configuration
 
     @property
-    def bugzoo(self) -> BugZooClient:
+    def environment(self) -> Environment:
         """
-        The BugZoo daemon.
+        A description of the simulated environment.
         """
-        return self._bugzoo
+        return self.__environment
 
     @property
-    def alive(self) -> bool:
+    def container(self) -> Container:
         """
-        A flag indicating whether or not this sandbox is alive.
+        The BugZoo container underlying this sandbox.
         """
-        # FIXME should also check that container is alive via BugZoo API
-        return self.__container is not None
+        return self.__container
 
-    def _start(self, mission: Mission) -> None:
+    def _start(self) -> None:
```

Comment:
Where is this called? I mean who instructs the sandbox to start the SITL?

Reply:
`_start` and `_stop` are called from outside of the sandbox instance. There are two ways that those methods will now be called:

1. a static method of `Sandbox` tentatively called `build` will implement a context manager that launches a sandbox inside a given Docker container and automatically starts and stops the SITL as the context is entered and exited.
   ```
   with Sandbox.build(bz, container, state, env, config) as sandbox:
      pass
   ```

2. the caller manually constructs a `Sandbox` instance and invokes `_start` and `_stop` at the appropriate time. This is useful for cases where context managers are clunky or inappropriate.